<a href="https://colab.research.google.com/github/KoeusIss/orvp/blob/main/0x01_fe.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Feature Engineering
This data science task is built on Kaggle competition released by **Optiver** to predict the realized volatility (market fluctuation).
In this step we try to explore some potential options to produce necessary features in order to maximize our chances finding a decent prediction.
To acheive this step we need some financial basics to understand our objective and the possible extra information we can extract from our datasets so we need to shed the light on 4 concept as discribed in the competition.
* bid/ask spread
* Weighted averaged price
* Log returns
* Realized volatilit

### bid/ask spread

As different stocks trade on different level on the market we take the ratio of best offer price and best bid price to calculate the bid-ask spread.

The formula of bid/ask spread can be written in below form:
$$BidAskSpread = BestOffer / BestBid - 1$$

And the spread _spread_:
$$ Spread = AskPrice -  BidPrice$$

### Weighted averaged price

The order book is also one of the primary source for stock valuation. A fair book-based valuation must take two factors into account: the level and the size of orders. In this competition we used weighted averaged price, or WAP, to calculate the instantaneous stock valuation and calculate realized volatility as our target.

The formula of WAP can be written as below, which takes the top level price and volume information into account:

$$𝑊𝐴𝑃=\frac{𝐵𝑖𝑑𝑃𝑟𝑖𝑐𝑒1∗𝐴𝑠𝑘𝑆𝑖𝑧𝑒1+𝐴𝑠𝑘𝑃𝑟𝑖𝑐𝑒1∗𝐵𝑖𝑑𝑆𝑖𝑧𝑒1}{𝐵𝑖𝑑𝑆𝑖𝑧𝑒1+𝐴𝑠𝑘𝑆𝑖𝑧𝑒1}$$
 
As you can see, if two books have both bid and ask offers on the same price level respectively, the one with more offers in place will generate a lower stock valuation, as there are more intended seller in the book, and more seller implies a fact of more supply on the market resulting in a lower stock valuation.

### Log returns
Returns are widely used in finance, however log returns are preferred whenever some mathematical modelling is required. Calling  $𝑆_𝑡$  the price of the stock  $𝑆$  at time  $𝑡$ , we can define the log return between  $𝑡_1$  and  $𝑡_2$  as:
$$𝑟_{𝑡1,𝑡2}=\log(𝑆_{𝑡2}, 𝑆_{𝑡1})$$
 
Usually, we look at log returns over fixed time intervals, so with 10-minute log return we mean  $𝑟_𝑡=𝑟_{𝑡−10 𝑚𝑖𝑛,𝑡}$ .

### Realized volatility
When we trade options, a valuable input to our models is the standard deviation of the stock log returns. The standard deviation will be different for log returns computed over longer or shorter intervals, for this reason it is usually normalized to a 1-year period and the annualized standard deviation is called **volatility**.
We will compute the log returns over all consecutive book updates and we define the realized volatility,  $\sigma$ , as the squared root of the sum of squared log returns. $$\sigma=\sqrt{\Sigma_t r^2_{t-1,t}}$$

In [1]:
# # Connect with GoogleDrive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [12]:
# Import libraries

import numpy as np
import pandas as pd
import matplotlib as mpl
import matplotlib.pyplot as plt
from joblib import Parallel, delayed
import seaborn as sns

# Constant
BASE_PATH = "/content/drive/MyDrive/DS/optiver-realized-volatility-prediction/"
TEST_FILE = BASE_PATH + "test.csv"
TRAIN_FILE = BASE_PATH + "train.csv"

In [4]:
# Financial utilities functions

def wap(df, index):
    """
    Compute the Waighted Avergae Price
    """
    bid_price = df[f"bid_price{index}"]
    bid_size = df[f"bid_size{index}"]
    ask_price = df[f"ask_price{index}"]
    ask_size = df[f"ask_size{index}"]
    return bid_price * ask_size + ask_price * bid_size / (bid_size + ask_size)

def log_return(series):
    """
    Compute the log_return of WAP
    """
    return np.log(series).diff()

def realized_volatility(series):
    """
    Compute the realized volatility
    """
    return np.sqrt(np.sum(series**2))

def total_volume(df):
    """
    Compute the total volume between asks and bids
    """
    ask_size_1 = df["ask_size1"]
    ask_size_2 = df["ask_size2"]
    bid_size_1 = df["bid_size1"]
    bid_size_2 = df["bid_size2"]
    return ask_size_1 + ask_size_2 + bid_size_1 + bid_size_2

def imbalance_volume(df):
    """
    Computes the imbalance volume
    """
    ask_size_1 = df["ask_size1"]
    ask_size_2 = df["ask_size2"]
    bid_size_1 = df["bid_size1"]
    bid_size_2 = df["bid_size2"]
    return np.abs(ask_size_1 + ask_size_2 - bid_size_1 - bid_size_2)

def spread_price(df, index):
    """
    Compute the spread
    """
    ask_price = df[f"ask_price{index}"]
    bid_price = df[f"bid_price{index}"]
    return ask_price - bid_price

def bid_ask_spread(df, index):
    """
    Compute the bid/ask spread
    """
    spread_price = df[f"spread_price{index}"]
    bid_price = df[f"bid_price{index}"]
    return spread_price / bid_price

In [5]:
def set_book_features(df, window_size=100):
    """
    Set book features // Windowing rows on columns
    """
    # First level aggregation
    df["wap_1"]                = wap(df, 1)
    df["wap_2"]                = wap(df, 2)
    df["log_return_1"]         = df.groupby("time_id")["wap_1"].apply(log_return)
    df["log_return_2"]         = df.groupby("time_id")["wap_2"].apply(log_return)
    df["spread_price_1"]       = spread_price(df, 1)
    df["spread_price_2"]       = spread_price(df, 2)
    df["bid_ask_spread_1"]     = spread_price(df, 1)
    df["bid_ask_spread_2"]     = spread_price(df, 2)
    df['wap_balance']          = np.abs(df['wap_1'] - df['wap_2'])
    df["total_volume"]         = total_volume(df)
    df["imbalance_volume"]     = imbalance_volume(df)
    
    # Second level aggregation
    aggregation_dict = {
        "log_return_1": [realized_volatility],
        "log_return_2": [realized_volatility],
        "wap_1": [np.mean, np.std],
        "wap_2": [np.mean, np.std],
        "bid_price1": [np.mean, np.max],
        "bid_price2": [np.mean, np.max],
        "ask_price1": [np.mean, np.min],
        "ask_price2": [np.mean, np.min],
        "bid_size1": [np.mean, np.sum],
        "bid_size2": [np.mean, np.sum],
        "ask_size1": [np.mean, np.sum],
        "ask_size2": [np.mean, np.sum],
        "spread_price_1": [np.mean, np.std],
        "spread_price_2": [np.mean, np.std],
        "bid_ask_spread_1": [np.mean, np.std],
        "bid_ask_spread_2": [np.mean, np.std],
        "wap_balance": [np.mean, np.std],
        "total_volume": [np.mean, np.std],
        "imbalance_volume": [np.mean, np.std]
    }
    features_lst = []
    for second in range(0, 600, window_size):
        df_feature = df[df['seconds_in_bucket'] >= second].groupby("time_id").agg(aggregation_dict)
        df_feature.columns = ['_'.join(col) + f"_s{600 - second}" for col in df_feature.columns]
        features_lst.append(df_feature)
    return pd.concat(features_lst, axis=1)

In [6]:
def set_trade_features(df, window_size=100):
    """
    Set trade features // Windowing rows on columns
    """
    # First level aggregation
    df["log_return"]   = df.groupby("time_id")["price"].apply(log_return)
    df['amount']       = df['price'] * df['size']
    
    # Second level aggregation
    aggregation_dict = {
        "log_return": [realized_volatility],
        "size": [np.mean, np.sum],
        "amount": [np.mean, np.sum],
        "order_count": [np.mean, np.sum],
    }
    features_lst = []
    for second in range(0, 600, window_size):
        df_feature = df[df['seconds_in_bucket'] >= second].groupby("time_id").agg(aggregation_dict)
        df_feature.columns = ['_'.join(col) + f"_s{600 - second}" for col in df_feature.columns]
        features_lst.append(df_feature)
    return pd.concat(features_lst, axis=1)

After setting the new features we tried to convert the time series datasets to supervised problem by windowing data by seconds_in_bucket so instead of 600 tic of seconds we regroup each 100 seconds and biuld a second level of aggregation.

In [7]:
def preprocess(ids_list, mode, window_size=60):
    """
    Preprocess dataset
    """
    def create_stock_df(stock_id):
        book = pd.read_parquet(f"{BASE_PATH}book_{mode}.parquet/stock_id={stock_id}")
        trade = pd.read_parquet(f"{BASE_PATH}trade_{mode}.parquet/stock_id={stock_id}")

        book_features = set_book_features(book, window_size)
        trade_features = set_trade_features(trade, window_size)
        
        features = book_features.join(trade_features, how='outer')
        features = features.reset_index()
        features["stock_id"] = stock_id
        return features
      
    df_list = Parallel(n_jobs=-1, verbose=1)(
        delayed(create_stock_df)(_id) for _id in ids_list
    )
    return pd.concat(df_list, ignore_index = True)

In [14]:
# Get the train file
train = pd.read_csv(TRAIN_FILE)
# For rapid and simple demonstration lets take only 5 stock_id
train = train[train["stock_id"].isin([0, 1, 2, 3, 4])]
train_ids = train.stock_id.unique()

We use paralllel computing distribution to fasten the load process

In [16]:
# Preprocess data + FE
train_df = preprocess(train_ids, "train")

[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.
/usr/local/lib/python3.7/dist-packages/joblib/externals/loky/process_executor.py:691: UserWarning: A worker stopped while some jobs were given to the executor. This can be caused by a too short worker timeout or by a memory leak.
  "timeout or by a memory leak.", UserWarning
[Parallel(n_jobs=-1)]: Done   5 out of   5 | elapsed:  3.8min finished


In [17]:
train_df

time_id  ...  stock_id
0            5  ...         0
1           11  ...         0
2           16  ...         0
3           31  ...         0
4           62  ...         0
...        ...  ...       ...
19145    32751  ...         4
19146    32753  ...         4
19147    32758  ...         4
19148    32763  ...         4
19149    32767  ...         4

[19150 rows x 432 columns]

Instead of 8 features we built a 432 feature describing a 6 window of time with statistics and financial features

In [18]:
# Merge and select
df_train = train.merge(train_df, on=["stock_id", "time_id"], how="left")

In [20]:
# let's check the missing data
df_train.isna().sum()

stock_id                                   0
time_id                                    0
target                                     0
log_return_1_realized_volatility_s600      0
log_return_2_realized_volatility_s600      0
                                        ... 
size_sum_s60                             722
amount_mean_s60                          722
amount_sum_s60                           722
order_count_mean_s60                     722
order_count_sum_s60                      722
Length: 433, dtype: int64

In [21]:
# Time windowing produce some missing value when there's no sufficient data in the last bucket seconds
# Forward filling the missing
df_train = df_train.ffill()

In [22]:
df_train

stock_id  time_id  ...  order_count_mean_s60  order_count_sum_s60
0             0        5  ...              2.750000                 22.0
1             0       11  ...              2.666667                  8.0
2             0       16  ...              2.500000                  5.0
3             0       31  ...              5.000000                 10.0
4             0       62  ...              5.500000                 11.0
...         ...      ...  ...                   ...                  ...
19145         4    32751  ...              7.250000                 29.0
19146         4    32753  ...              1.000000                  3.0
19147         4    32758  ...              3.400000                 17.0
19148         4    32763  ...              8.500000                 17.0
19149         4    32767  ...              1.000000                  2.0

[19150 rows x 433 columns]